### PR TITLE
ENH: `__array_function__` support for `np.fft` and `np.linalg`

### PR DESCRIPTION
**Pull request by shoyer**
xref #12028

### CODE REVIEW

**Review by hameerabbasi — APPROVED** [review]
This one (particularly `linalg`) in particular will be useful for PyData/Sparse long term. Thanks for this.

I see that ignoring `None` is a theme here, so I'm going to ignore it completely in future reviews. I'm sure you carefully considered the pros and cons of such an approach.

File: numpy/fft/fftpack.py [hunk 1]
```diff
@@ -101,6 +102,11 @@ def _unitary(norm):
     return norm is not None
 
 
+def _fft_dispatcher(a, n=None, axis=None, norm=None):
```
Comment (mhvk) [inline numpy/fft/fftpack.py]: As far as I can see, in this whole file, there are only two signatures: `(a, n, axis, norm)`, and `(a, s, axes, norm)` - I think in such a well-constructed file, it is a lot clearer to define the two signatures once and use them throughout.

File: numpy/fft/helper.py [hunk 2]
```diff
@@ -19,6 +20,11 @@
 integer_types = integer_types + (integer,)
 
 
+def _fftshift_dispatcher(x, axes=None):
```
Comment (mhvk) [inline numpy/fft/helper.py]: I would again suggest just using a single dispatcher.

**Review by mhvk — COMMENTED** [review]
Apart from my suggestion to make an exception for fft to the rule that every function has its own dispatcher, this looks good.

**Comment by charris** [comment]
This looks to be still under discussion. Ping when ready.

**Comment by charris** [comment]
Remove WIP from title when ready.

**Comment by shoyer** [comment]
Yeah, I got a little carried away with copy & paste here. This should be much less repetitive now.

**Comment by mhvk** [comment]
Well, for `fft` at least, it is clear what the best option is!  I'm quite happy for this to go in.

**Pull request merged**